x-pack/solutions/security/plugins/security_solution/server/endpoint/services/agent/clients/microsoft_defender_endpoint/microsoft_defender_endpoint_agent_status_client.ts

```diff
@@ -0,0 +1,117 @@
+import type { AgentStatusRecords } from '../../../../../../common/endpoint/types';
+import { HostStatus } from '../../../../../../common/endpoint/types';
+
+export class MicrosoftDefenderEndpointAgentStatusClient extends AgentStatusClient {
+  protected readonly agentType: ResponseActionAgentType = 'microsoft_defender_endpoint';
+
+  protected readonly connectorActions: NormalizedExternalConnectorClient;
+
+  constructor(options: AgentStatusClientOptions) {
+    super(options);
+
+    if (!options.connectorActionsClient) {
+      throw new AgentStatusClientError(
+        'connectorActionsClient is required to create an instance of MicrosoftDefenderEndpointAgentStatusClient'
+      );
+    }
+
+    this.connectorActions = new NormalizedExternalConnectorClient(
+      options.connectorActionsClient,
+      this.log
+    );
+    this.connectorActions.setup(MICROSOFT_DEFENDER_ENDPOINT_CONNECTOR_ID);
+  }
+
+  protected getAgentStatusFromMachineHealthStatus(
+    healthStatus: MicrosoftDefenderEndpointMachine['healthStatus'] | undefined
+  ): HostStatus {
+    // Definition of sensor health status can be found here:
+    // https://learn.microsoft.com/en-us/defender-endpoint/check-sensor-status
+
+    switch (healthStatus) {
+      case 'Active':
+        return HostStatus.HEALTHY;
+
+      case 'Inactive':
+        return HostStatus.INACTIVE;
+
+      case 'ImpairedCommunication':
+      case 'NoSensorData':
+      case 'NoSensorDataImpairedCommunication':
+        return HostStatus.UNHEALTHY;
+
+      default:
+        return HostStatus.UNENROLLED;
+    }
+  }
+
+  public async getAgentStatuses(agentIds: string[]): Promise<AgentStatusRecords> {
+    const esClient = this.options.esClient;
+    const metadataService = this.options.endpointService.getEndpointMetadataService();
+
+    try {
+      const [{ data: msMachineListResponse }, allPendingActions] = await Promise.all([
+        this.connectorActions.execute<MicrosoftDefenderEndpointAgentListResponse>({
+          params: {
+            subAction: MICROSOFT_DEFENDER_ENDPOINT_SUB_ACTION.GET_AGENT_LIST,
+            subActionParams: { id: agentIds },
+          },
+        }),
+
+        // Fetch pending actions summary
+        getPendingActionsSummary(esClient, metadataService, this.log, agentIds),
+      ]);
+
+      const machinesById = keyBy(msMachineListResponse?.value ?? [], 'id');
+      const pendingActionsByAgentId = keyBy(allPendingActions, 'agent_id');
+
+      return agentIds.reduce<AgentStatusRecords>((acc, agentId) => {
+        const thisMachine = machinesById[agentId];
+        const thisAgentPendingActions = pendingActionsByAgentId[agentId];
+
+        acc[agentId] = {
+          agentId,
+          agentType: this.agentType,
+          found: !!thisMachine,
+          // Unfortunately, it does not look like MS Defender has a way to determine
+          // if a host is isolated or not via API, so we just set this to false
```

Comment:
I see Microsoft's API is even better than CrowdStrikes ;)

I did some digging, and wondering if this workaround could help us:

1. Use the List machineActions API to retrieve all actions performed on devices. You can filter the results to focus on isolation actions.
2. To check the isolation status of a specific device, filter the actions by the device's ID.
3. Check the retrieved actions to determine the device's isolation status. If the most recent action is an 'Isolate' action with a status of 'Succeeded', the device is currently isolated. If the most recent action is an 'Unisolate' action with a status of 'Succeeded', the device is not isolated.

I know it's quite cumbersome, but getting the status is pretty crucial functionality, so might be worth it?

Reply:
I thought about that, and don't think its a good approach. This API on our side allows for multiple agents IDs to be passed, so we would have to do that processing on each individual agent, which would be time consuming and we would be hitting the external API allot. There are also edge cases with failures or for hosts that may be offline that would further complicate the approach. 

I will continue to search for a solution to see if I can find out. I know for sure that the API MS Defender uses to support their UI returns a property called `IsolationState` - but I have not yet been able to find a public API that returns that

Reply:
Fine by me, thanks!

Reply:
So... I have been searching for a while now to see if there is a more reliable way to get a hosts's isolation status from MS and I can't seem to identify one - pretty hard to believe that its this difficult to get.

I'm going to implement what you suggested above - I also found some posts online that alluded to this being the only way to do it.  I'll push that in a follow up PR.

Thanks for the review and suggestion